theano/tensor/opt.py

```diff
@@ -3253,6 +3253,50 @@ def local_flatten_lift(node):
 ##################
 
 
+@register_canonicalize
+@register_specialize
+@gof.local_optimizer([T.Reshape])
+def local_useless_reshape(node):
+    """ Reshape{ndim=1}(vector) -> vector
+
+    Remove Reshape when both the input and the output have a single
+    dimension.
+    If the shape of the input can't be inferred, an Assert is inserted to
+    ensure that the given new shape is the same as the original one.
+    """
+    if isinstance(node.op, T.Reshape):
+        x, shape = node.inputs
+        if node.op.ndim == 1 and x.ndim == 1:
+            out = x
+
+            new_shape_len = None
+            try:
+                new_shape_len = get_scalar_constant_value(shape.shape[0])
+            except NotScalarConstantError:
+                pass
+            if new_shape_len != 1:
+                out = assert_op(out, T.eq(shape.shape[0], 1))
+
+            try:
+                new_shape = get_scalar_constant_value(shape[0])
+                if new_shape == -1:
+                    return [out]
+
+                if new_shape == 1:
+                    out = T.addbroadcast(out, 0)
```

Comment:
The thing is, the output of reshape can have a different broadcast pattern :
```
x = theano.tensor.vector('x')
y = x.reshape([1])
```
`y` broadcast is `(True,)`, whereas `x` is `(False,)`.
Should I just return and leave the original Reshape in this case ?

If I don't do this, I get an error because I replaced Reshape with a node whose output has different type.